File: results/1.0/GIF/gif-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct {
    HParser *signature;
    HParser *version;
    HParser *header;
    HParser *screen_desc;
    HParser *global_color_table;
    HParser *extension;
    HParser *image_block;
    HParser *trailer;
    HParser *gif;
} GIFParsers;

GIFParsers init_gif_parsers() {
    GIFParsers p;
    
    // Header Block
    p.signature = h_token((uint8_t*)"GIF", 3);
    p.version = h_choice(h_token((uint8_t*)"87a", 3),
                        h_token((uint8_t*)"89a", 3),
                        NULL);
    p.header = h_sequence(p.signature, p.version, NULL);

    // Logical Screen Descriptor
    HParser *screen_width = h_uint16();
    HParser *screen_height = h_uint16();
    HParser *packed_field = h_bits(8, false);
    HParser *bg_color_index = h_uint8();
    HParser *pixel_aspect_ratio = h_uint8();
    p.screen_desc = h_sequence(screen_width, screen_height, packed_field,
                              bg_color_index, pixel_aspect_ratio, NULL);

    // Global Color Table
    HParser *rgb = h_repeat_n(h_uint8(), 3);
    p.global_color_table = h_many(rgb);

    // Extension Blocks
    HParser *extension_introducer = h_ch(0x21);
    
    // Graphics Control Extension
    HParser *gce_label = h_ch(0xF9);
    HParser *gce_block_size = h_ch(4);
    HParser *gce_packed = h_bits(8, false);
    HParser *delay_time = h_uint16();
    HParser *transparent_idx = h_uint8();
    HParser *gce = h_sequence(extension_introducer, gce_label, gce_block_size,
                             gce_packed, delay_time, transparent_idx, h_ch(0), NULL);

    // Comment Extension
    HParser *comment_label = h_ch(0xFE);
    HParser *comment_data = h_many1(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    HParser *comment = h_sequence(extension_introducer, comment_label, comment_data,
                                 h_ch(0), NULL);

    // Plain Text Extension
    HParser *text_label = h_ch(0x01);
    HParser *text_block_size = h_ch(12);
    HParser *text_grid = h_repeat_n(h_uint16(), 4);
    HParser *cell_size = h_repeat_n(h_uint8(), 2);
    HParser *text_colors = h_repeat_n(h_uint8(), 2);
    HParser *text_data = h_many1(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    HParser *plain_text = h_sequence(extension_introducer, text_label, text_block_size,
                                    text_grid, cell_size, text_colors, text_data,
                                    h_ch(0), NULL);

    // Application Extension
    HParser *app_label = h_ch(0xFF);
    HParser *app_block_size = h_ch(11);
    HParser *app_id = h_repeat_n(h_uint8(), 8);
    HParser *auth_code = h_repeat_n(h_uint8(), 3);
    HParser *app_data = h_many1(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    HParser *application = h_sequence(extension_introducer, app_label, app_block_size,
                                     app_id, auth_code, app_data, h_ch(0), NULL);

    p.extension = h_choice(gce, comment, plain_text, application, NULL);

    // Image Block
    HParser *image_separator = h_ch(0x2C);
    HParser *position = h_repeat_n(h_uint16(), 2);
    HParser *dimensions = h_repeat_n(h_uint16(), 2);
    HParser *image_packed = h_bits(8, false);
    HParser *local_color_table = h_many(rgb);
    HParser *lzw_min_code = h_uint8();
    HParser *image_data = h_many1(h_sequence(h_uint8(), h_many(h_uint8()), NULL));
    p.image_block = h_sequence(image_separator, position, dimensions, image_packed,
                              local_color_table, lzw_min_code, image_data,
                              h_ch(0), NULL);

    // Trailer
    p.trailer = h_ch(0x3B);

    // Complete GIF
    p.gif = h_sequence(p.header, p.screen_desc, p.global_color_table,
                      h_many(h_choice(p.extension, p.image_block, NULL)),
                      p.trailer, NULL);

    return p;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <gif_file>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t *data = malloc(size);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(f);
        return 1;
    }

    if (fread(data, 1, size, f) != size) {
        perror("Failed to read file");
        free(data);
        fclose(f);
        return 1;
    }

    GIFParsers parsers = init_gif_parsers();
    HParseResult *result = h_parse(parsers.gif, data, size);

    if (result) {
        printf("Successfully parsed GIF file\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse GIF file\n");
    }

    free(data);
    fclose(f);
    return 0;
}

File: results/1.0/GIF/gif-claude-3-5-sonnet-20241022.rs
use nom::{
    bytes::complete::{tag, take as take_bytes},
    combinator::map,
    multi::many0,
    number::complete::{le_u16, le_u8},
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct GifHeader {
    signature: [u8; 3],
    version: [u8; 3],
}

#[derive(Debug)]
struct LogicalScreenDescriptor {
    width: u16,
    height: u16,
    global_color_table_flag: bool,
    color_resolution: u8,
    sort_flag: bool,
    size_of_global_color_table: u8,
    background_color_index: u8,
    pixel_aspect_ratio: u8,
}

#[derive(Debug)]
struct RGB {
    r: u8,
    g: u8,
    b: u8,
}

#[derive(Debug)]
struct ImageDescriptor {
    left_position: u16,
    top_position: u16,
    width: u16,
    height: u16,
    local_color_table_flag: bool,
    interlace_flag: bool,
    sort_flag: bool,
    size_of_local_color_table: u8,
}

#[derive(Debug)]
struct GraphicControlExtension {
    disposal_method: u8,
    user_input_flag: bool,
    transparent_color_flag: bool,
    delay_time: u16,
    transparent_color_index: u8,
}

#[derive(Debug)]
struct CommentExtension {
    comment_data: Vec<u8>,
}

#[derive(Debug)]
struct PlainTextExtension {
    text_grid_left_position: u16,
    text_grid_top_position: u16,
    text_grid_width: u16,
    text_grid_height: u16,
    character_cell_width: u8,
    character_cell_height: u8,
    text_foreground_color_index: u8,
    text_background_color_index: u8,
    plain_text_data: Vec<u8>,
}

#[derive(Debug)]
struct ApplicationExtension {
    application_identifier: [u8; 8],
    authentication_code: [u8; 3],
    application_data: Vec<u8>,
}

#[derive(Debug)]
struct GifData {
    header: GifHeader,
    logical_screen_descriptor: LogicalScreenDescriptor,
    global_color_table: Option<Vec<RGB>>,
    blocks: Vec<Block>,
}

#[derive(Debug)]
enum Block {
    Image {
        descriptor: ImageDescriptor,
        local_color_table: Option<Vec<RGB>>,
        image_data: Vec<u8>,
    },
    GraphicControl(GraphicControlExtension),
    Comment(CommentExtension),
    PlainText(PlainTextExtension),
    Application(ApplicationExtension),
}

fn parse_header(input: &[u8]) -> IResult<&[u8], GifHeader> {
    let (input, (signature, version)) = tuple((
        take_bytes(3usize),
        take_bytes(3usize),
    ))(input)?;

    Ok((input, GifHeader {
        signature: signature.try_into().unwrap(),
        version: version.try_into().unwrap(),
    }))
}

fn parse_packed_field(input: &[u8]) -> IResult<&[u8], (bool, u8, bool, u8)> {
    let (input, packed) = le_u8(input)?;
    Ok((input, (
        (packed & 0x80) != 0,
        ((packed & 0x70) >> 4) as u8,
        (packed & 0x08) != 0,
        (packed & 0x07) as u8,
    )))
}

fn parse_logical_screen_descriptor(input: &[u8]) -> IResult<&[u8], LogicalScreenDescriptor> {
    let (input, (width, height)) = tuple((le_u16, le_u16))(input)?;
    let (input, (global_color_table_flag, color_resolution, sort_flag, size_of_global_color_table)) = parse_packed_field(input)?;
    let (input, (background_color_index, pixel_aspect_ratio)) = tuple((le_u8, le_u8))(input)?;

    Ok((input, LogicalScreenDescriptor {
        width,
        height,
        global_color_table_flag,
        color_resolution,
        sort_flag,
        size_of_global_color_table,
        background_color_index,
        pixel_aspect_ratio,
    }))
}

fn parse_rgb(input: &[u8]) -> IResult<&[u8], RGB> {
    let (input, (r, g, b)) = tuple((le_u8, le_u8, le_u8))(input)?;
    Ok((input, RGB { r, g, b }))
}

fn parse_color_table(size: u8) -> impl Fn(&[u8]) -> IResult<&[u8], Vec<RGB>> {
    move |input| {
        let count = 1 << (size + 1);
        let mut colors = Vec::with_capacity(count as usize);
        let mut remaining = input;

        for _ in 0..count {
            let (new_input, color) = parse_rgb(remaining)?;
            colors.push(color);
            remaining = new_input;
        }

        Ok((remaining, colors))
    }
}

fn parse_sub_blocks(input: &[u8]) -> IResult<&[u8], Vec<u8>> {
    let mut data = Vec::new();
    let mut remaining = input;

    loop {
        let (new_input, block_size) = le_u8(remaining)?;
        if block_size == 0 {
            return Ok((new_input, data));
        }

        let (new_input, block_data) = take_bytes(block_size as usize)(new_input)?;
        data.extend_from_slice(block_data);
        remaining = new_input;
    }
}

fn parse_image_descriptor(input: &[u8]) -> IResult<&[u8], ImageDescriptor> {
    let (input, _) = tag(&[0x2C])(input)?;
    let (input, (left_position, top_position, width, height)) = 
        tuple((le_u16, le_u16, le_u16, le_u16))(input)?;
    let (input, packed) = le_u8(input)?;

    Ok((input, ImageDescriptor {
        left_position,
        top_position,
        width,
        height,
        local_color_table_flag: (packed & 0x80) != 0,
        interlace_flag: (packed & 0x40) != 0,
        sort_flag: (packed & 0x20) != 0,
        size_of_local_color_table: packed & 0x07,
    }))
}

fn parse_graphic_control_extension(input: &[u8]) -> IResult<&[u8], GraphicControlExtension> {
    let (input, _) = tuple((tag(&[0x21]), tag(&[0xF9]), tag(&[0x04])))(input)?;
    let (input, packed) = le_u8(input)?;
    let (input, (delay_time, transparent_color_index)) = tuple((le_u16, le_u8))(input)?;
    let (input, _) = tag(&[0x00])(input)?;

    Ok((input, GraphicControlExtension {
        disposal_method: (packed & 0x1C) >> 2,
        user_input_flag: (packed & 0x02) != 0,
        transparent_color_flag: (packed & 0x01) != 0,
        delay_time,
        transparent_color_index,
    }))
}

fn parse_comment_extension(input: &[u8]) -> IResult<&[u8], CommentExtension> {
    let (input, _) = tuple((tag(&[0x21]), tag(&[0xFE])))(input)?;
    let (input, comment_data) = parse_sub_blocks(input)?;

    Ok((input, CommentExtension { comment_data }))
}

fn parse_plain_text_extension(input: &[u8]) -> IResult<&[u8], PlainTextExtension> {
    let (input, _) = tuple((tag(&[0x21]), tag(&[0x01]), tag(&[0x0C])))(input)?;
    let (input, (
        text_grid_left_position,
        text_grid_top_position,
        text_grid_width,
        text_grid_height,
        character_cell_width,
        character_cell_height,
        text_foreground_color_index,
        text_background_color_index
    )) = tuple((
        le_u16, le_u16, le_u16, le_u16,
        le_u8, le_u8, le_u8, le_u8
    ))(input)?;
    let (input, plain_text_data) = parse_sub_blocks(input)?;

    Ok((input, PlainTextExtension {
        text_grid_left_position,
        text_grid_top_position,
        text_grid_width,
        text_grid_height,
        character_cell_width,
        character_cell_height,
        text_foreground_color_index,
        text_background_color_index,
        plain_text_data,
    }))
}

fn parse_application_extension(input: &[u8]) -> IResult<&[u8], ApplicationExtension> {
    let (input, _) = tuple((tag(&[0x21]), tag(&[0xFF]), tag(&[0x0B])))(input)?;
    let (input, identifier) = take_bytes(8usize)(input)?;
    let (input, auth_code) = take_bytes(3usize)(input)?;
    let (input, application_data) = parse_sub_blocks(input)?;

    Ok((input, ApplicationExtension {
        application_identifier: identifier.try_into().unwrap(),
        authentication_code: auth_code.try_into().unwrap(),
        application_data,
    }))
}

fn parse_block(input: &[u8]) -> IResult<&[u8], Block> {
    let (input, peek_byte) = take_bytes(1usize)(input)?;
    match peek_byte[0] {
        0x2C => {
            let (input, descriptor) = parse_image_descriptor(input)?;
            let (input, local_color_table) = if descriptor.local_color_table_flag {
                let (input, table) = parse_color_table(descriptor.size_of_local_color_table)(input)?;
                (input, Some(table))
            } else {
                (input, None)
            };
            let (input, _) = le_u8(input)?; // LZW minimum code size
            let (input, image_data) = parse_sub_blocks(input)?;
            Ok((input, Block::Image {
                descriptor,
                local_color_table,
                image_data,
            }))
        },
        0x21 => {
            let (_, extension_type) = take_bytes(2usize)(input)?;
            match extension_type[1] {
                0xF9 => map(parse_graphic_control_extension, Block::GraphicControl)(input),
                0xFE => map(parse_comment_extension, Block::Comment)(input),
                0x01 => map(parse_plain_text_extension, Block::PlainText)(input),
                0xFF => map(parse_application_extension, Block::Application)(input),
                _ => panic!("Unknown extension type"),
            }
        },
        _ => panic!("Unknown block type"),
    }
}

fn parse_gif(input: &[u8]) -> IResult<&[u8], GifData> {
    let (input, header) = parse_header(input)?;
    let (input, logical_screen_descriptor) = parse_logical_screen_descriptor(input)?;
    
    let (input, global_color_table) = if logical_screen_descriptor.global_color_table_flag {
        let (input, table) = parse_color_table(logical_screen_descriptor.size_of_global_color_table)(input)?;
        (input, Some(table))
    } else {
        (input, None)
    };

    let (input, blocks) = many0(parse_block)(input)?;
    let (input, _) = tag(&[0x3B])(input)?; // Trailer

    Ok((input, GifData {
        header,
        logical_screen_descriptor,
        global_color_table,
        blocks,
    }))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <gif_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1]).expect("Failed to open file");
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).expect("Failed to read file");

    match parse_gif(&buffer) {
        Ok((_, gif_data)) => println!("{:#?}", gif_data),
        Err(e) => eprintln!("Error parsing GIF: {:?}", e),
    }
}